[WIP] MPLUGINTESTING-93

aamotharald opened the pull request:
starting some documentation activities now and performing testing of the JUnit5 maven-plugin testing capabilities with existing open source maven plugins.

@gnodet I have made a small change in the MojoExtension as I think it's valid to have no configuration for your maven plugin at all.

slachiewicz commented on maven-plugin-testing-harness/src/main/java/org/apache/maven/api/plugin/testing/InjectMojo.java:
```diff
@@ -16,7 +16,7 @@
  * under the License.
  */
-package org.apache.maven.plugin.testing.junit5;
+package org.apache.maven.api.plugin.testing;
```
this may collide with Maven 4 apis

aamotharald replied:
Hi @slachiewicz ,
yes this might collide and this I am doing by intent.
Consider the following use case:
Somebody uses `@InjectMojo`on maven version 3 and creates JUnit Jupiter test cases to test his/her Mojo.

Then he/she decides to migrate to Maven4.
As the codelines differ between maven3/maven4, there might be glitches / issues. But the migration will be a lot less cumbersome if you don't have to rename all the imports just because the package changes from maven3 to maven4. This is why I would like to have the same Packages on Maven3/4.
Does that make sense or should we strictly separate the packages and make the maven3-4 migration cumbersome for the consumers?

gnodet commented:
This is really interesting.
I'd like to see some changes from https://github.com/apache/maven-plugin-testing/commit/c6ef221ab068ee487e1c535aa2b8925ecc4f0db6 back ported to make the annotations identical (the `@Inherited` and the changes in `InjectMojo`).
Maybe also the new `@Basedir` annotation ?

aamotharald commented:
Hi @gnodet ,
I have been able to to make the Annotations identical (added @Inherited and removed the boolean flag from `InjectMojo`).
I was also able to backport the @Basedir Annotation and added a unit test for that behaviour.

So what's left open is the stubbing. I made a check and backporting all the stubs seems to be a bigger task. My current opinion is to NOT offer the stubs on maven 3 (which means maven-plugin-testing with stubs is only possible with old JUnit 4 style tests extending the AbstractMojoTestCase). What is your opinion there?
